### PR TITLE
Resampling is now a separate, streamable filter

### LINKED ISSUE #46: Make image resampling a streamable filter

### PR DESCRIPTION
**Pull request by dzenanz**
Closes #46.

### CODE REVIEW

**Comment by dzenanz** [comment]
This crashes on Windows for me in [WSACleanup()](https://github.com/InsightSoftwareConsortium/ITK/blob/e54719632a51dd7aee23683ceecc6d514cff0c28/Modules/ThirdParty/KWSys/src/KWSys/SystemInformation.cxx#L5334).

File: include/itkTileMerging.h [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+#ifndef itkTileMerging_h
+#define itkTileMerging_h
```
Comment (thewtex) [inline include/itkTileMerging.h]: -> `itk::TileMergeImageFilter` following ITK naming conventions.

File: include/itkTileMerging.h [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+#include "itkTileMontage.h"
```
Comment (thewtex) [inline include/itkTileMerging.h]: -> `itk::TileMontageImageFilter` following ITK naming conventions.

File: include/itkTileMerging.h [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+#include "itkLinearInterpolateImageFunction.h"
+#include "itkSimpleFastMutexLock.h"
+
+namespace itk
+{
+/** \class TileMerging
+ * \brief Resamples an n-Dimensional mosaic of images into a single big image.
```
Comment (thewtex) [inline include/itkTileMerging.h]: big -> composite

File: include/itkTileMerging.h [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+ *
+ * CropToFill indicates whether the big image will be cropped so it
```
Comment (thewtex) [inline include/itkTileMerging.h]: big -> composite

File: include/itkTileMerging.h [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+ * entirely consists of input tiles (no default background filling).
+ * If CropToFill is false, the big image will have the extent to include
```
Comment (thewtex) [inline include/itkTileMerging.h]: big -> composite

File: include/itkTileMerging.h [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+ * all of the input tiles. The pixels not covered by any input tile
+ * will have the value specified by the Background member variable.
```
Comment (thewtex) [inline include/itkTileMerging.h]: `Background` -> `BackgroundValue` to be more explicit and follow ITK naming conventions

File: include/itkTileMerging.h [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  /** Run-time type information (and related methods). */
+  itkTypeMacro(TileMerging, TileMontage);
+
+  /** Dimensionality of input images. */
+  itkStaticConstMacro(ImageDimension, unsigned int, ImageType::ImageDimension);
+
+  /** This is envisioned to be the primary way of setting inputs.
+   * All required inputs are taken from TileMontage. Alternatively,
+   * inherited members can be called individually, e.g.:
+   * SetMontageSize(), SetInputTile(), SetOriginAdjustment() etc. */
```
Comment (thewtex) [inline include/itkTileMerging.h]: Interesting. Do we need to inherit also?

File: include/itkTileMerging.h [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+  using typename Superclass::ImageIndexType;
+
+  /**  Type for the transform. */
+  using TransformType = typename Superclass::PCMType::TransformType;
+  using TransformPointer = typename TransformType::Pointer;
+  using TransformConstPointer = typename TransformType::ConstPointer;
+
+  ///** Type for the output: Using Decorator pattern for enabling
+  //*  the Transform to be passed in the data pipeline */
+  //using TransformOutputType = DataObjectDecorator< TransformType >;
```
Comment (thewtex) [inline include/itkTileMerging.h]: remove

File: include/itkTileMerging.h [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+  ///** Smart Pointer type to a DataObject. */
+  //using DataObjectPointer = typename DataObject::Pointer;
```
Comment (thewtex) [inline include/itkTileMerging.h]: remove

File: include/itkTileMerging.h [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+  ImageType * GetOutput(unsigned int idx);
+
+  /** Set size of the image mosaic. */
+  void SetMontageSize(SizeType montageSize);
+
+  /** To be called for each tile position in the mosaic
+   * before the call to Update(). */
+  void SetInputTile(TileIndexType position, ImageType* image)
+  {
+    SizeValueType linInd = this->nDIndexToLinearIndex(position);
```
Comment (thewtex) [inline include/itkTileMerging.h]: `linInd` -> `linearIndex`

Comment (thewtex) [reply include/itkTileMerging.h]: 'const`

File: include/itkTileMerging.h [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+   * and can be run in parallel with other indices. */
+  void ResampleSingleRegion(SizeValueType regionIndex);
+
+private:
+    bool      m_CropToFill; //crop to avoid backfround filling?
+    PixelType m_Background; //default background value (not covered by any input tile)
+
+    std::vector<TransformConstPointer> m_Transforms;
+    std::vector<SimpleFastMutexLock>   m_TileReadLocks; //to avoid reading the same tile by more than one thread in parallel
+    std::vector<ImagePointer>          m_Metadata; // images with empty buffers
```
Comment (thewtex) [inline include/itkTileMerging.h]: Why is this called `Metadata`? What are the empty buffers for? How are they populated?

File: include/itkTileMerging.hxx [hunk 12]
```diff
@@ -0,0 +1,590 @@
...
+}
+
+template <typename TImageType, typename TInterpolator>
+void
+TileMerging<TImageType, TInterpolator>
+::SplitRegionAndCopyContributions(
+    std::vector<RegionType>& regions,
+    std::vector<ContributingTiles>& regionContributors,
+    RegionType newRegion,
+    size_t ori, //oldRegionIndex
```
Comment (thewtex) [inline include/itkTileMerging.hxx]: use `oldRegionIndex`

File: include/itkTileMerging.hxx [hunk 13]
```diff
@@ -0,0 +1,590 @@
...
+    SizeValueType tileIndex)
+{
+  for (int d = ImageDimension - 1; d >= 0; d--)
+    {
+    SizeValueType nrSize = newRegion.GetSize(d);
```
Comment (thewtex) [inline include/itkTileMerging.hxx]: use `newRegion` and `oldRegion`

File: include/itkTileMerging.hxx [hunk 14]
```diff
@@ -0,0 +1,590 @@
...
+void
+TileMerging<TImageType, TInterpolator>
+::GenerateData()
+{
+  ImagePointer outputImage = this->GetOutput();
+  RegionType reqR = outputImage->GetRequestedRegion();
+  outputImage->SetBufferedRegion(reqR);
+  outputImage->Allocate(false);
+
+  ////for debugging purposes, just color the regions by their contributing tiles
```
Comment (thewtex) [inline include/itkTileMerging.hxx]: Make this a flag (off by default) that can set on the filter and uncomment

File: include/itkTileMerging.hxx [hunk 15]
```diff
@@ -0,0 +1,590 @@
...
+      while (!oIt.IsAtEnd())
+        {
+        ContinuousIndexType continuousIndex = oIt.GetIndex();
+        continuousIndex += continuousIndexDifference;
+        oIt.Set(interp->EvaluateAtContinuousIndex(continuousIndex));
+        ++oIt;
+        }
+      }
+    }
+  else //more than one tile contributes
```
Comment (thewtex) [inline include/itkTileMerging.hxx]: Does this work when the output image region is split for streaming?

Comment (dzenanz) [reply include/itkTileMerging.hxx]: Yes

**Comment by dzenanz** [comment]
The existing test already uses streaming: `w->SetNumberOfStreamDivisions(3);` This is invoked near the end of itkMontageTestHelper.hxx (around line 200).

Working on rest of suggestions.

File: include/itkTileMerging.h (hunk 7 above)
Comment (dzenanz) [reply include/itkTileMerging.h]: We reduce a lot of duplication by inheriting.

File: include/itkTileMerging.h (hunk 11 above)
Comment (dzenanz) [reply include/itkTileMerging.h]: Initially it was just metadata. Later I used them for storing images (possibly just required region). I will rename.

**Comment by dzenanz** [comment]
Now off to the class/file renames.

**Pull request merged**